Approved/3-multidimensional-arrays.md

```diff
@@ -0,0 +1,701 @@
+Each axis of a multidimensional arrays can be sliced by _either_ a value of type `Range` or a value of type `Int`; for example, `(Int, Int)`, `(Range, Int)`, `(Int, Range)`, and `(Range, Range)` are valid subscripts for a value of type `[|'T|]`.
+As shown in Example 5 below, for each `Int` in an index tuple, the dimensionality (aka rank) of the array is reduced by one.
+That is, indexing a `[|'T|]` by `(Range, Range)` returns a rank-2 array (`[|'T|]`), while indexing by `(Int, Range)` or `(Range, Int)` returns an ordinary rank-1 array (`[|'T|]`).
+Just as with indices like `(Int, Int)` and `(Int, Int, Int)`, subscripts that return slices can also be used in copy-and-replace expressions, as shown in Example 6.
+When using `Range` values to index one or more axes in a multidimensional array, `...` is shorthand the `Range` value `0..1..(n - 1)` where `n` is the length of the axes being indexed.
+
+When used in `for` loops, multidimensional arrays iterate "on the left," yielding loop variables of one rank lower than the array being looped over, as shown in Example 7, below.
+
+Finally, to support multidimensional arrays, this proposal also suggests extending the `Microsoft.Quantum.Arrays` namespace with the following functions that can be used to implement libraries for working with multidimensional arrays:
+
+- `internal function AsNDArray2<'TElement>(data : ['TElement], strides : [Int], offset : Int, shape : [Int]) : [|'TElement|]`
+- `internal function AsNDArray3<'TElement>(data : ['TElement], strides : [Int], offset : Int, shape : [Int]) : [||'TElement||]`
+- `internal function AsNDArray4<'TElement>(data : ['TElement], strides : [Int], offset : Int, shape : [Int]) : [|||'TElement|||]`
+- `internal function AsNDArray5<'TElement>(data : ['TElement], strides : [Int], offset : Int, shape : [Int]) : [||||'TElement||||]`
+- `internal function AsNDArray6<'TElement>(data : ['TElement], strides : [Int], offset : Int, shape : [Int]) : [|||||'TElement|||||]`
+- `internal function AsNDArray7<'TElement>(data : ['TElement], strides : [Int], offset : Int, shape : [Int]) : [||||||'TElement||||||]`
+- `internal function AsNDArray8<'TElement>(data : ['TElement], strides : [Int], offset : Int, shape : [Int]) : [|||||||'TElement|||||||]`
+- `internal function NDArrayData<'TElement, 'TArray>(data : 'TArray) : ['TElement]`
+- `internal function NDArrayStrides<'TArray>(data : 'TArray) : [Int]`
+- `internal function NDArrayShape<'TArray>(data : 'TArray) : [Int]`
+- `internal function NDArrayOffset<'TArray>(data : 'TArray) : Int`
+
+Each of these five functions would be `body intrinsic;`, and together form the contract between the runtime and the core Q# libraries needed to support this proposed feature (see Example 7, below). By necessity, these functions are "unsafe," in that direct access to these functions would allow violating invariants of multidimensional arrays or bypass the type system to expose runtime failures, necessitating the `internal` modifier.
```

Comment:
There are more than 5 functions listed here, and I'd like to find a way to not have to support AsNDArray* as intrinsic. I'd also suggest to rename `NDArrayShape` to `NDArraySize`.

Reply:
My thought was that the return types being different in each case would require separate callables, but we could apply the same trick of using unsafe type parameters:

```
internal function UnsafeAsNDArray<'TElement, 'TArray>(data : ['TElement], strides : [Int], offset : Int, shape : [Int]) : 'TArray {
    body intrinsic;
}

internal function AsNDArray2<'TElement>(data : ['TElement], strides : [Int], offset : Int, shape : [Int]) : [|'TElement|] {
    return UnsafeAsNDArray<'TElement, [|'TElement|]>(data, strides, offset, shape);
}
```